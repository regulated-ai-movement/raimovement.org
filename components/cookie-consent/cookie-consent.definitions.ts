export const COOKIE_CONSENT_KEY = '__cookie-consent'

export enum Consent {
  Accepted = 'accepted',
  Declined = 'declined'
}

export enum ConsentLabel {
  Accept = 'Accept',
  Decline = 'Decline'
}
