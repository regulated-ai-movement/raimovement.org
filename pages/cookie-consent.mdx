# Cookie Consent Notice

Our website, raimovement.org, respects your privacy.

While we do not actively use cookies ourselves to track your visit to our website, our site does make use of Google Analytics, a service which transmits website traffic data to Google servers.

## What is a cookie?

Cookies are small text files that are placed on your machine to help the site provide a better user experience. In general, cookies are used to retain user preferences, store information for things like shopping carts, and provide anonymized tracking data to third party applications like Google Analytics.

## What does Google Analytics do?

Google Analytics is a web analysis service provided by Google Inc. (“Google”). Google utilizes the Data collected to track and examine the use of this website, to prepare reports on its activities and share them with other Google services. Google may use the Data collected to contextualize and personalize the ads of its own advertising network.

This data is not tied to personally identifiable information. We do not use this data to identify users personally. Our purpose in using analytics is to understand how well our site is working, how people are finding it, and what content is being viewed. This allows us to improve the service we provide to you.
Consent

By continuing to navigate our website, you hereby acknowledge and agree to raimovement.org's use of cookies. If you do not wish to accept the use of Google Analytics cookies, you can disable them by adjusting your browser settings or using Google’s opt-out tool.

## More information

If you wish to learn more about cookies and how to manage or delete them, we recommend visiting www.aboutcookies.org.

For more information on how Google uses data when you use our site, please visit: www.google.com/policies/privacy/partners/

Please also review our [Privacy Policy](./privacy-policy.mdx) to understand how we collect and use data.

Thank you for visiting raimovement.org! We appreciate your trust in us.