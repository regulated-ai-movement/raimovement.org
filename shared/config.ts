export enum NavigationIconLinks {
  LinkedIn = 'https://www.linkedin.com/company/regulated-ai-movement/',
  Twitter = 'https://twitter.com/raimovement',
  GitHub = 'https://github.com/regulated-ai-movement/raimovement.org',
  Discord = 'https://discord.gg/tckpz2scJw'
}

export enum StaticRoutes {
  CookieConsent = 'cookie-consent',
  PrivacyPolicy = 'privacy-policy'
}
