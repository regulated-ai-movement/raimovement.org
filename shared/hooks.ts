import { useState } from "react";

const useLocalStorage = <T>(key: string, initialValue: unknown): [T, (value: T) => void] => {
    const [storedValue, setStoredValue] = useState(() => {
        try {
            const item = window?.localStorage?.getItem(key);
            return item ? JSON.parse(item) : initialValue;
        } catch (error) {
            console.log(error);
            return initialValue;
        }
    });

    const setValue = (value: T) => {
        try {
            const valueToStore =
                value instanceof Function ? value(storedValue) : value;
            setStoredValue(valueToStore);
            window?.localStorage?.setItem(key, JSON.stringify(valueToStore));
        } catch (error) {
            console.log(error);
        }
    };

    return [storedValue, setValue];
}

export { useLocalStorage }