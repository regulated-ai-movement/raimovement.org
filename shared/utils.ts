export const isSSR = (): Boolean => typeof window === 'undefined'
export const isCrawler = (): Boolean => navigator.userAgent.includes('Crawler')
export const isClient = (): Boolean => !isSSR() && !isCrawler()