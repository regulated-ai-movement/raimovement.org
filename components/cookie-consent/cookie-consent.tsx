import { StaticRoutes } from '../../shared/config'
import { getCookie, setCookie } from 'nextjs-cookie'
import { useCallback, useEffect, useState } from 'react'
import { COOKIE_CONSENT_KEY, Consent } from './cookie-consent.definitions'
import { updateConsent } from '../../shared/gtag'
import { usePathname } from 'next/navigation'

export const CookieConsent = (): JSX.Element | null => {
  const [cookieConsent, setCookieConsent] = useState<boolean>(true)
  const pathname = usePathname()
  const hideConsent =
    pathname.includes(StaticRoutes.PrivacyPolicy) ||
    pathname.includes(StaticRoutes.CookieConsent)

  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
      const consentValue = (event.target as HTMLButtonElement).value as Consent
      setCookie(COOKIE_CONSENT_KEY, consentValue)
      setCookieConsent(true)
      if (consentValue === Consent.Accepted) {
        updateConsent()
      }
    },
    []
  )

  useEffect(() => {
    const cookieConsent = getCookie(COOKIE_CONSENT_KEY)
    if (cookieConsent) {
      setCookieConsent(true)
    } else {
      setCookieConsent(false)
    }
  }, [])

  if (cookieConsent || hideConsent) {
    return null
  }

  return (
    <div className="fixed top-0 left-0 right-0 bottom-0 flex items-center justify-center bg-theme-blue bg-opacity-80">
      <div className="shadow-lg bg-theme-white text-theme-blue text-center p-10 h-auto">
        <p>
          {
            'Welcome ðŸ‘‹! Is it ok if we use cookies to improve your experience? '
          }
          <a className="underline" href={StaticRoutes.CookieConsent}>
            {'Read more'}
          </a>
        </p>
        <div className="flex justify-center gap-2 mt-4">
          <button
            className="p-2 m-2 bg-green-600 text-theme-white rounded-md"
            onClick={handleClick}
            value={Consent.Accepted}
          >
            {'Accept'}
          </button>
          <button
            className="p-2 m-2 border-solid border-2 border-theme-red text-theme-red rounded-md"
            onClick={handleClick}
            value={Consent.Declined}
          >
            {'Decline'}
          </button>
        </div>
      </div>
    </div>
  )
}
