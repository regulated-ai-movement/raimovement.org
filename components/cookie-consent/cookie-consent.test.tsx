import { CookieConsent } from './cookie-consent'
import { ConsentLabel } from './cookie-consent.definitions'
import { fireEvent, render, screen } from '@testing-library/react'
import { vi } from 'vitest'
import { setCookie } from 'nextjs-cookie'
import { updateConsent } from '../../shared/gtag'
import { StaticRoutes } from '../../shared/config'
import { usePathname } from 'next/navigation'

vi.mock('nextjs-cookie', () => ({
  getCookie: vi.fn(),
  setCookie: vi.fn()
}))

vi.mock('../../shared/gtag', () => ({
  updateConsent: vi.fn()
}))

vi.mock('next/navigation', () => ({
  usePathname: vi.fn(() => 'other-route')
}))

describe('CookieConsent', () => {
  beforeAll(() => {
    vi.clearAllMocks()
  })

  test('should render', () => {
    const wrapper = render(<CookieConsent />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should render with accept button', () => {
    render(<CookieConsent />)

    expect(screen.getByText(ConsentLabel.Accept)).toBeVisible()
    expect(screen.getByText(ConsentLabel.Decline)).toBeVisible()
  })

  test('should disappear on click and set cookie', () => {
    render(<CookieConsent />)

    const acceptButton = screen.getByText(ConsentLabel.Accept)

    fireEvent.click(acceptButton)

    expect(acceptButton).not.toBeVisible()
    expect(setCookie).toHaveBeenCalled()
    expect(updateConsent).toHaveBeenCalled()
  })

  test('should disappear on privacy and terms pages', () => {
    vi.mocked(usePathname).mockReturnValue(`/${StaticRoutes.CookieConsent}`)

    render(<CookieConsent />)

    const acceptButton = screen.queryByText(ConsentLabel.Accept)

    expect(acceptButton).toBeNull()
  })
})
