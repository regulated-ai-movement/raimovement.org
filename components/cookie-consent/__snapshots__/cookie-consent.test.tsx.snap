// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CookieConsent > should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fixed top-0 left-0 right-0 bottom-0 flex items-center justify-center bg-theme-blue bg-opacity-80"
      >
        <div
          class="shadow-lg bg-theme-white text-theme-blue text-center p-10 h-auto"
        >
          <p>
            Welcome ðŸ‘‹! Is it ok if we use cookies to improve your experience? 
            <a
              class="underline"
              href="cookie-consent"
            >
              Read more
            </a>
          </p>
          <div
            class="flex justify-center gap-2 mt-4"
          >
            <button
              class="p-2 m-2 bg-green-600 text-theme-white rounded-md"
              value="accepted"
            >
              Accept
            </button>
            <button
              class="p-2 m-2 border-solid border-2 border-theme-red text-theme-red rounded-md"
              value="declined"
            >
              Decline
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="fixed top-0 left-0 right-0 bottom-0 flex items-center justify-center bg-theme-blue bg-opacity-80"
    >
      <div
        class="shadow-lg bg-theme-white text-theme-blue text-center p-10 h-auto"
      >
        <p>
          Welcome ðŸ‘‹! Is it ok if we use cookies to improve your experience? 
          <a
            class="underline"
            href="cookie-consent"
          >
            Read more
          </a>
        </p>
        <div
          class="flex justify-center gap-2 mt-4"
        >
          <button
            class="p-2 m-2 bg-green-600 text-theme-white rounded-md"
            value="accepted"
          >
            Accept
          </button>
          <button
            class="p-2 m-2 border-solid border-2 border-theme-red text-theme-red rounded-md"
            value="declined"
          >
            Decline
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
